gateway-server/src/main/java/org/apache/knox/gateway/websockets/ProxyWebSocketAdapter.java

```diff
@@ -198,12 +232,29 @@ public void onError(Throwable cause) {
 
       @Override
       public void onMessageText(String message, Object session) {
+        LOG.logMessage("[From Backend <---]" + message);
+        remoteLock.lock();
         final RemoteEndpoint remote = getRemote();
```

Comment:
So I don't know this code, but it seems weird to me that getRemote returns null at all. It looks like the remote null check is fixing the symptom not the root cause. Is it possible to make getRemote block and never return null? I think that would avoid having to buffer messages at all?

Reply:
What I observed that remote  was being created in other thread which was executing onWebSocketConnect() while this current thread received a message. This might be an limitation with Jetty itself, check: https://github.com/signalapp/Signal-Desktop/issues/1840. 

I checked jetty examples and even there the server is using locks before accessing remote, as mentioned in the above link.

Reply:
I tried adding a fixed delay of few seconds in onMessageText() API but I was still seeing the issue in some cases. So adding a fixed configurable delay will not work.